source/common/config/incremental_subscription_impl.h

```diff
@@ -0,0 +1,312 @@
+#include "common/common/backoff_strategy.h"
+#include "common/common/logger.h"
+#include "common/common/token_bucket_impl.h"
+#include "common/config/utility.h"
+#include "common/grpc/common.h"
+#include "common/protobuf/protobuf.h"
+#include "common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * TODO SOMETHING SOMETHING subscription. Also handles per-xDS API stats/logging.
+ */
+template <class ResourceType>
+class IncrementalSubscriptionImpl
+    : public IncrementalSubscription<ResourceType>,
+      Grpc::TypedAsyncStreamCallbacks<envoy::api::v2::IncrementalDiscoveryResponse>,
+      Logger::Loggable<Logger::Id::config> {
+public:
+  IncrementalSubscriptionImpl(const LocalInfo::LocalInfo& local_info,
+                              Grpc::AsyncClientPtr async_client, Event::Dispatcher& dispatcher,
+                              const Protobuf::MethodDescriptor& service_method,
+                              Runtime::RandomGenerator& random,
+                              const RateLimitSettings& rate_limit_settings,
+                              IncrementalSubscriptionStats stats,
+                              ControlPlaneStats control_plane_stats)
+      : type_url_(Grpc::Common::typeUrl(ResourceType().GetDescriptor()->full_name())),
+        local_info_(local_info), async_client_(std::move(async_client)),
+        service_method_(service_method), rate_limiting_enabled_(rate_limit_settings.enabled_),
+        stats_(stats), control_plane_stats_(control_plane_stats) {
+    retry_timer_ = dispatcher.createTimer([this]() -> void { establishNewStream(); });
+    if (rate_limiting_enabled_) {
+      // Default Bucket contains 100 tokens maximum and refills at 10 tokens/sec.
+      limit_request_ = std::make_unique<TokenBucketImpl>(
+          rate_limit_settings.max_tokens_, dispatcher.timeSystem(), rate_limit_settings.fill_rate_);
+      drain_request_timer_ = dispatcher.createTimer([this]() -> void { drainRequests(); });
+    }
+    backoff_strategy_ =
+        std::make_unique<JitteredBackOffStrategy>(50,    // TODO RETRY_INITIAL_DELAY_MS,
+                                                  30000, // TODO RETRY_MAX_DELAY_MS,
+                                                  random);
+    establishNewStream();
+  }
+
+  // GRPC ACTUAL STREAM HANDLING STUFF
+  //  void start() { establishNewStream(); }
+
+  void setRetryTimer() {
+    retry_timer_->enableTimer(std::chrono::milliseconds(backoff_strategy_->nextBackOffMs()));
+  }
+
+  void establishNewStream() {
+    ENVOY_LOG(debug, "Establishing new gRPC bidi stream for {}", service_method_.DebugString());
+    stream_ = async_client_->start(service_method_, *this);
+    if (stream_ == nullptr) {
+      ENVOY_LOG(warn, "Unable to establish new stream");
+      onIncrementalConfigFailed(nullptr);
+      setRetryTimer();
+      return;
+    }
+
+    // "must be populated for first request in a stream"
+    request_.clear_initial_resource_versions();
```

Comment:
We should probably clear all of `request_` on a new stream.

Reply:
Done.